File: backend/app/services/routing_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.models import Worker
from math import radians, sin, cos, sqrt, atan2

# Narrow street graph (same as before)
NARROW_GRAPH = {
    "nodes": {
        "worker_home": {"lat": 12.9716, "lng": 77.5946},
        "junction1": {"lat": 12.9718, "lng": 77.5948},
        "bin_b1": {"lat": 12.9718, "lng": 77.5948},
        "junction2": {"lat": 12.9720, "lng": 77.5950},
        "bin_b2": {"lat": 12.9722, "lng": 77.5952},
        "narrow_start": {"lat": 12.9719, "lng": 77.5949},
        "narrow_end": {"lat": 12.9721, "lng": 77.5951},
    },
    "edges": {
        "worker_home->junction1": {"distance": 0.05, "type": "main"},
        "junction1->bin_b1": {"distance": 0.02, "type": "narrow"},
        "junction1->junction2": {"distance": 0.08, "type": "main"},
        "junction2->bin_b2": {"distance": 0.03, "type": "main"},
        "junction2->narrow_start": {"distance": 0.01, "type": "alley"},
        "narrow_start->narrow_end": {"distance": 0.02, "type": "narrow"},
        "narrow_end->bin_b1": {"distance": 0.01, "type": "narrow"},
    }
}

def dijkstra(start_id: str, end_id: str):
    nodes = list(NARROW_GRAPH["nodes"].keys())
    distances = {node: float('inf') for node in nodes}
    previous = {node: None for node in nodes}
    distances[start_id] = 0
    unvisited = set(nodes)

    while unvisited:
        current = min(unvisited, key=lambda node: distances[node])
        if distances[current] == float('inf'):
            break
        unvisited.remove(current)
        if current == end_id:
            break
        neighbors = []
        edges = NARROW_GRAPH["edges"]
        for edge in edges:
            parts = edge.split("->")
            if parts[0] == current and parts[1] in unvisited:
                neighbors.append(parts[1])
            elif parts[1] == current and parts[0] in unvisited:
                neighbors.append(parts[0])
        for neighbor in neighbors:
            # find edge distance
            edge_key = f"{current}->{neighbor}" if f"{current}->{neighbor}" in edges else f"{neighbor}->{current}"
            alt = distances[current] + edges[edge_key]["distance"]
            if alt < distances[neighbor]:
                distances[neighbor] = alt
                previous[neighbor] = current

    path = []
    node = end_id
    while node is not None:
        path.insert(0, node)
        node = previous[node]
    if path[0] != start_id:
        return []
    return path

def get_route(worker_lat: float, worker_lng: float, bin_lat: float, bin_lng: float):
    nodes = NARROW_GRAPH["nodes"]
    # Find closest start node
    start_id = min(nodes.keys(), key=lambda nid: (nodes[nid]["lat"] - worker_lat)**2 + (nodes[nid]["lng"] - worker_lng)**2)
    end_id = min(nodes.keys(), key=lambda nid: (nodes[nid]["lat"] - bin_lat)**2 + (nodes[nid]["lng"] - bin_lng)**2)
    path_ids = dijkstra(start_id, end_id)
    route_coords = [nodes[nid] for nid in path_ids]
    return {"path_ids": path_ids, "route_coords": route_coords}

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.models import Worker
from math import radians, sin, cos, sqrt, atan2

async def find_nearest_worker(bin_lat: float, bin_lng: float, db: AsyncSession):
    result = await db.execute(select(Worker).where(Worker.status == "active"))
    workers = result.scalars().all()
    if not workers:
        return None
    
    def haversine(lat1, lon1, lat2, lon2):
        R = 6371
        dlat = radians(lat2 - lat1)
        dlon = radians(lon2 - lon1)
        a = sin(dlat/2)**2 + cos(radians(lat1)) * cos(radians(lat2)) * sin(dlon/2)**2
        c = 2 * atan2(sqrt(a), sqrt(1-a))
        return R * c
    
    nearest = min(workers, key=lambda w: haversine(bin_lat, bin_lng, w.lat, w.lng))
    return nearest

File: frontend/services/routingService.ts
import { narrowGraph } from './graphData';

type GraphNodes = typeof narrowGraph.nodes;
type GraphEdges = typeof narrowGraph.edges;

// Helper: get all node IDs
const getAllNodeIds = (): string[] => Object.keys(narrowGraph.nodes);

// Find shortest path using Dijkstra's algorithm
export const findShortestPath = (startId: string, endId: string): string[] => {
  const nodes = getAllNodeIds();
  const distances: Record<string, number> = {};
  const previous: Record<string, string | null> = {};
  const unvisited = new Set<string>(nodes);

  // Initialize distances
  for (const node of nodes) {
    distances[node] = Infinity;
    previous[node] = null;
  }
  distances[startId] = 0;

  // Cast edges to an indexable type for safe access
  const edges = narrowGraph.edges as Record<string, { distance: number; type: string }>;

  while (unvisited.size > 0) {
    // Find unvisited node with smallest distance
    let current: string | null = null;
    for (const node of unvisited) {
      if (current === null || distances[node] < distances[current]) {
        current = node;
      }
    }
    if (current === null || distances[current] === Infinity) break;
    if (current === endId) break;

    unvisited.delete(current);

    // Find neighbours
    const neighbours: string[] = [];
    for (const edge of Object.keys(edges)) {
      const parts = edge.split('->');
      if (parts[0] === current && unvisited.has(parts[1])) {
        neighbours.push(parts[1]);
      } else if (parts[1] === current && unvisited.has(parts[0])) {
        neighbours.push(parts[0]);
      }
    }

    for (const neighbour of neighbours) {
      // Determine edge key (always "from->to")
      let edgeKey: string;
      if (edges[`${current}->${neighbour}`]) {
        edgeKey = `${current}->${neighbour}`;
      } else {
        edgeKey = `${neighbour}->${current}`;
      }
      const alt = distances[current] + edges[edgeKey].distance;
      if (alt < distances[neighbour]) {
        distances[neighbour] = alt;
        previous[neighbour] = current;
      }
    }
  }

  // Reconstruct path
  const path: string[] = [];
  let curr: string | null = endId;
  while (curr !== null) {
    path.unshift(curr);
    curr = previous[curr];
  }
  if (path[0] !== startId) return []; // no valid route
  return path;
};

// Get route coordinates from worker location to bin location
export const getRoute = (
  workerLat: number,
  workerLng: number,
  binLat: number,
  binLng: number
): { pathIds: string[]; routeCoords: { lat: number; lng: number }[] } => {
  const nodes = narrowGraph.nodes;
  let startNodeId = 'worker_home';
  let endNodeId = 'bin_b1';

  // Find closest graph node to worker location
  let minStartDist = Infinity;
  for (const [id, coords] of Object.entries(nodes)) {
    const nodeCoords = coords as { lat: number; lng: number };
    const dist = Math.hypot(nodeCoords.lat - workerLat, nodeCoords.lng - workerLng);
    if (dist < minStartDist) {
      minStartDist = dist;
      startNodeId = id;
    }
  }

  // Find closest graph node to bin location
  let minEndDist = Infinity;
  for (const [id, coords] of Object.entries(nodes)) {
    const nodeCoords = coords as { lat: number; lng: number };
    const dist = Math.hypot(nodeCoords.lat - binLat, nodeCoords.lng - binLng);
    if (dist < minEndDist) {
      minEndDist = dist;
      endNodeId = id;
    }
  }

  const pathIds = findShortestPath(startNodeId, endNodeId);
  const routeCoords = pathIds.map(id => nodes[id as keyof typeof nodes] as { lat: number; lng: number });

  return { pathIds, routeCoords };
};